### DISCUSSION TITLE
OSError: [Errno 30] Read-only file system

### DISCUSSION THREAD

**Discussion by Stawa** [vercel/vercel#8734] [Help]
Hello, I have one problem. When I tried to open and read a file but it didn't work, it only gives me an error about `OSError: [Errno 30] Read-only file system`. I have no problems with my current code when I tested it out using a local host. I am not sure what the problem is, and I don't know what to do. Any solutions?

![image](https://user-images.githubusercontent.com/69102292/196139072-648884eb-a68e-489b-89d2-a6e2b50437e0.png)

**Comment by amyegan** [comment]
Thanks for reaching out @Stawa. It looks like you're trying to save a file from your serverless function, but app files become read-only when deployed. So this code probably works in your local dev environment, but not when built for production and deployed.

If you only need to store the file temporarily, you can use the `/tmp` of your Serverless Function. If you need more permanent storage, I would recommend using a database or other permanent storage option.

**Reply by Stawa** [reply]
That makes sense. I chose `/tmp` for my current project since I don't want to keep the file forever. Do you have any examples of its methods?

**Reply by amyegan** [reply]
Sorry for the very late response! We updated the ["How can I use files in Serverless Functions on Vercel?"](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions) guide with a section about [using temporary storage](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions#using-temporary-storage). It includes some example code for temporarily writing a PDF in a serverless function.
